exchanges/gemini/gemini_websocket.go

```diff
@@ -411,60 +545,51 @@ func (g *Gemini) wsProcessUpdate(result WsMarketUpdateResponse, pair currency.Pa
 			return
 		}
 	} else {
-		var asks, bids []orderbook.Item
-		var trades []trade.Data
-		for i := range result.Events {
-			switch result.Events[i].Type {
-			case "trade":
-				tSide, err := order.StringToOrderSide(result.Events[i].MakerSide)
-				if err != nil {
-					g.Websocket.DataHandler <- order.ClassificationError{
-						Exchange: g.Name,
-						Err:      err,
-					}
-				}
-				trades = append(trades, trade.Data{
-					Timestamp:    time.Unix(0, result.TimestampMS*int64(time.Millisecond)),
-					CurrencyPair: pair,
-					AssetType:    asset.Spot,
-					Exchange:     g.Name,
-					Price:        result.Events[i].Price,
-					Amount:       result.Events[i].Amount,
-					Side:         tSide,
-					TID:          strconv.FormatInt(result.Events[i].ID, 10),
-				})
-			case "change":
-				item := orderbook.Item{
-					Amount: result.Events[i].Remaining,
-					Price:  result.Events[i].Price,
-				}
-				if strings.EqualFold(result.Events[i].Side, order.Ask.String()) {
-					asks = append(asks, item)
-				} else {
-					bids = append(bids, item)
-				}
-			default:
-				g.Websocket.DataHandler <- fmt.Errorf("%s - Unhandled websocket update: %+v", g.Name, result)
-			}
-		}
-		if len(trades) > 0 && g.IsSaveTradeDataEnabled() {
-			err := trade.AddTradesToBuffer(g.Name, trades...)
-			if err != nil {
-				g.Websocket.DataHandler <- err
-			}
-		}
 		if len(asks) == 0 && len(bids) == 0 {
 			return
 		}
 		err := g.Websocket.Orderbook.Update(&buffer.Update{
-			Asks:       asks,
-			Bids:       bids,
-			Pair:       pair,
-			UpdateTime: time.Unix(0, result.TimestampMS*int64(time.Millisecond)),
-			Asset:      asset.Spot,
+			Asks:  asks,
+			Bids:  bids,
+			Pair:  pair,
+			Asset: asset.Spot,
 		})
 		if err != nil {
-			g.Websocket.DataHandler <- fmt.Errorf("%v %v", g.Name, err)
+			g.Websocket.DataHandler <- err
+			return
+		}
+	}
+
+	if len(result.AuctionEvents) > 0 {
+		g.Websocket.DataHandler <- result.AuctionEvents
+	}
+
+	if !g.IsSaveTradeDataEnabled() {
+		return
+	}
+
+	var trades []trade.Data
+	for x := range result.Trades {
+		tSide, err := order.StringToOrderSide(result.Trades[x].Side)
+		if err != nil {
+			g.Websocket.DataHandler <- order.ClassificationError{
+				Exchange: g.Name,
+				Err:      err,
+			}
 		}
+		trades = append(trades, trade.Data{
+			Timestamp:    time.Unix(result.Trades[x].Timestamp/1000, 0),
+			CurrencyPair: pair,
+			AssetType:    asset.Spot,
+			Exchange:     g.Name,
+			Price:        result.Trades[x].Price,
+			Amount:       result.Trades[x].Quantity,
+			Side:         tSide,
+			TID:          strconv.FormatInt(result.Trades[x].EventID, 10),
```

Comment:
Please add a `!= "" && != "0"` check before setting the `TID` here. Not all trades seem to come with eventIDs and Gemini sends "0". The trade table has a unique check for TID + exchangename, so if multiple have "0", then it will be ignored/overwritten. For example, LINK-USD.
If the property `TID` is not set, then you can still proceed as there is another unique constrain that checks everything else but the ID for uniqueness:
```
 CONSTRAINT uniquetrade
        unique(exchange_name_id, base, quote, asset, price, amount, side, timestamp)
```

Reply:
Nice catch, this was caused by an invalid tag! Should be fixed now